mux/mux.go

```diff
@@ -3,104 +3,87 @@ package mux
 import (
 	"context"
 	"errors"
-	"net"
+	"log"
 	"net/http"
+	"strings"
 	"time"
 
-	"github.com/soheilhy/cmux"
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/h2c"
 	"google.golang.org/grpc"
 )
 
 const defaultGracePeriod = 5 * time.Second
+const grpcRequestMarker = "application/grpc"
 
-// Serve starts a TCP listener and serves the registered protocol
-// servers (i.e., http, grpc, etc.) and blocks until server exits.
-// Context can be cancelled to perform graceful shutdown.
-// Use WithHTTP() and WithGRPC() to register procol servers.
+// Serve starts a TCP listener and serves the registered protocol servers and blocks
+// until server exits. Context can be cancelled to perform graceful shutdown.
+// Use WithHTTP() and WithGRPC() to register protocol servers.
 func Serve(ctx context.Context, listenAddr string, opts ...Option) error {
-	var mux cmuxWrapper
+	var mux protocolMuxServer
```

Comment:
aren't we multiplexing grpc & http? its not a protocol we are multiplexing, but just another way to communicate over http 1 or 2. 
`protocolMuxServer` is bit misleading there, as we are just doing grpc vs http here, and not multiplexing protocols like `ftp`, `http`, `smtp` .. etc using this